<launch >
    <arg name="node_name" default="mrf_detector" />

    <!-- Topic and frame parameters -->
    <arg name="map_topic_name" default="/map" />
    <arg name="scan_topic_name" default="/scan" />
    <arg name="map_frame_name" default="map" />
    <arg name="scan_frame_name" default="laser_front" />

    <!-- If publish_classified_scans is true, the above scan messages will be published. -->
    <arg name="publish_classified_scans" default="true" />
    <!-- Debug messages -->
    <arg name="aligned_scan_mrf" default="/aligned_scan_mrf" />
    <!-- sensor_msgs::LaserScan -->
    <arg name="misaligned_scan_mrf" default="/misaligned_scan_mrf" />
    <!-- sensor_msgs::LaserScan -->
    <arg name="unknown_scan_mrf" default="/unknown_scan_mrf" />
    <!-- sensor_msgs::LaserScan -->

    <!-- Parameters related to the likelihood distributions
         These values must be positive. -->
    <arg name="normal_distribution_mean" default="0.0" />
    <arg name="normal_distribution_var" default="0.01" />
    <arg name="exponential_distribution_lambda" default="2.0" />
    <arg name="max_residual_error" default="1.0" />
    <arg name="residual_error_resolution" default="0.05" />

    <!-- Parameters related to the failure probability calculation -->
    <!-- These values must be positive. -->
    <arg name="min_valid_residual_errors_num" default="10" />
    <!-- If you want to reduce the computational cost,
         please set small values to max_residual_errors_num, max_lpb_computation_num, and sampling_num.
         In particular, max_residual_errors_num is significant. -->
    <arg name="max_residual_errors_num" default="250" />
    <arg name="max_lpb_computation_num" default="1000" />
    <arg name="sampling_num" default="1000" />
    <!-- These values must be included from 0 to 1. -->
    <arg name="misalignment_ratio_threshold" default="0.1" />
    <arg name="unknown_ratio_threshold" default="0.7" />

    <!-- Parameters related to the failure probability calculation
         This is a row major matrix and the sum of each row must be 1. -->
    <rosparam param="/$(arg node_name)/transition_probability_matrix"> [0.8, 0.0, 0.2, 0.0, 0.8, 0.2, 0.333333, 0.333333, 0.333333] </rosparam >




    <!-- Failure detector node -->
    <node name="$(arg node_name)" pkg="mrflocalization_detector" type="mrf_detector" output="screen">

        <param name="map_topic_name" value="$(arg map_topic_name)" type="string" />
        <param name="scan_topic_name" value="$(arg scan_topic_name)" type="string" />
        <param name="map_frame_name" value="$(arg map_frame_name)" type="string" />
        <param name="scan_frame_name" value="$(arg scan_frame_name)" type="string" />

        <param name="aligned_scan_mrf" value="$(arg aligned_scan_mrf)" type="string" />
        <param name="misaligned_scan_mrf" value="$(arg misaligned_scan_mrf)" type="string" />
        <param name="unknown_scan_mrf" value="$(arg unknown_scan_mrf)" type="string" />
        <param name="publish_classified_scans" value="$(arg publish_classified_scans)" type="bool" />

        <param name="normal_distribution_mean" value="$(arg normal_distribution_mean)" type="double" />
        <param name="normal_distribution_var" value="$(arg normal_distribution_var)" type="double" />
        <param name="exponential_distribution_lambda" value="$(arg exponential_distribution_lambda)" type="double" />
        <param name="max_residual_error" value="$(arg max_residual_error)" type="double" />
        <param name="residual_error_resolution" value="$(arg residual_error_resolution)" type="double" />
        <param name="min_valid_residual_errors_num" value="$(arg min_valid_residual_errors_num)" type="int" />
        <param name="max_residual_errors_num" value="$(arg max_residual_errors_num)" type="int" />
        <param name="max_lpb_computation_num" value="$(arg max_lpb_computation_num)" type="int" />
        <param name="sampling_num" value="$(arg sampling_num)" type="int" />
        <param name="misalignment_ratio_threshold" value="$(arg misalignment_ratio_threshold)" type="double" />
        <param name="unknown_ratio_threshold" value="$(arg unknown_ratio_threshold)" type="double" />

    </node >
</launch >
